adapters/teads/teads.go

```diff
@@ -0,0 +1,178 @@
+				bannerCopy.W = &imp.Banner.Format[0].W
+				imp.Banner = &bannerCopy
+			}
+		}
+
+		var defaultImpExt DefaultBidderImpExtension
+		if err := json.Unmarshal(imp.Ext, &defaultImpExt); err != nil {
+			return &errortypes.BadInput{
+				Message: "Error parsing Imp.Ext object",
+			}
+		}
+		imp.TagID = strconv.Itoa(defaultImpExt.Bidder.PlacementId)
+		teadsImpExt := &TeadsImpExtension{
+			KV: TeadsKV{
+				PlacementId: defaultImpExt.Bidder.PlacementId,
+			},
+		}
+		if extJson, err := json.Marshal(teadsImpExt); err != nil {
+			return &errortypes.BadInput{
+				Message: "Error stringify Imp.Ext object",
+			}
+		} else {
+			imp.Ext = extJson
+		}
+	}
+	return nil
+}
+
+// Builds enpoint url based on adapter-specific pub settings from imp.ext
+func (a *TeadsAdapter) buildEndpointURL() (string, error) {
+	endpointParams := macros.EndpointTemplateParams{}
+	host, err := macros.ResolveMacros(a.endpointTemplate, endpointParams)
+
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: "Unable to parse endpoint url template: " + err.Error(),
+		}
+	}
+
+	thisURI, err := url.Parse(host)
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: "Malformed URL: " + err.Error(),
+		}
+	}
+
+	return thisURI.String(), nil
+}
+
+func (a *TeadsAdapter) MakeBids(internalRequest *openrtb2.BidRequest, _ *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(bidResp.SeatBid))
+
+	for _, sb := range bidResp.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+
+			var bidExtTeads TeadsBidExt
+			if err := json.Unmarshal(bid.Ext, &bidExtTeads); err != nil {
+				return nil, []error{err}
+			}
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid: &bid,
+				BidMeta: &openrtb_ext.ExtBidPrebidMeta{
+					RendererName:    bidExtTeads.Prebid.Meta.RendererName,
+					RendererVersion: bidExtTeads.Prebid.Meta.RendererVersion,
+				},
+				BidType: getMediaTypeForImp(bid.ImpID, internalRequest.Imp),
+			})
+		}
+	}
+	if bidResp.Cur != "" {
+		bidResponse.Currency = bidResp.Cur
+	}
+	return bidResponse, []error{}
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Video != nil {
```

Comment:
@github-maxime-liege please take a look at above comment suggesting to use `MType` field

Reply:
I took this piece of code on other adapters because it was needed, there is so many adapters that uses this code. As we are not setting the MType field in our adServer currently this could be our workaround for now.

Reply:
> I took this piece of code on other adapters because it was needed, there is so many adapters that uses this code. As we are not setting the MType field in our adServer currently this could be our workaround for now.

Thank you @github-maxime-liege  for update. 

Also on  - `I took this piece of code on other adapters because it was needed, there is so many adapters that uses this code`, 

The `Mtype` field was added in openrtb 2.6 spec and some of adapter predates before  2.6 spec  release. Therefore, few piece of code on other adapters might not be using `Mtype`. But going forward prebid team encourages `Mtype` field for such usage. 

Since `Teads` adServer is not able to set `MType` field. So this workaround will work now but only for non multi-format imps. Therefore should update https://github.com/prebid/prebid.github.io/pull/4871 doc PR to be explicit that multi format imps are not supported

Reply:
i updated the documentation according to your comment, so it is now clear for publishers that we will not support multiformat bids currently